### PR TITLE
refactor(textbox): Add the Lightweight Styling Text Box Style

### LINKED ISSUE #988: [Material] Lightweight Styling - TextBox Styles
Following the guidelines from https://github.com/unoplatform/Uno.Themes/issues/854, refactor the resources for all styles related to this control into ThemeDictionaries ("Default"/"Light") and make sure to reference these resources using the `ThemeResource` binding in ALL places within the style

### PR DESCRIPTION
**Pull request by nguyeenvy**
closes #988 

﻿GitHub Issue: #988

## PR Type
- Refactoring (no functional changes, no api changes)

## Description
TextBox.xaml has been modified. New resources have been created for overriding in other pages.
Filled
image
image

Outlined
image
image

Locally overridden with:

	<Page.Resources>
		
		<SolidColorBrush x:Key="FilledTextBoxBackground"
						 Color="Red" />
		<SolidColorBrush x:Key="FilledTextBoxBackgroundPointerOver"
						 Color="Orange" />
		<SolidColorBrush x:Key="FilledTextBoxBackgroundFocused"
						 Color="Yellow" />
		<SolidColorBrush x:Key="FilledTextBoxBackgroundDisabled"
						 Color="Green" />

		<SolidColorBrush x:Key="FilledTextBoxBorderBrush"
						 Color="Blue" />
		<SolidColorBrush x:Key="FilledTextBoxBorderBrushPointerOver"
						 Color="Indigo" />
		<SolidColorBrush x:Key="FilledTextBoxBorderBrushFocused"
						 Color="Pink" />
		<SolidColorBrush x:Key="FilledTextBoxBorderBrushDisabled"
						 Color="Red" />

		<SolidColorBrush x:Key="FilledTextBoxDeleteButtonForeground"
						 Color="Orange" />
		<SolidColorBrush x:Key="FilledTextBoxDeleteButtonForegroundPointerOver"
						 Color="Yellow" />
		<SolidColorBrush x:Key="FilledTextBoxDeleteButtonForegroundFocused"
						 Color="Green" />
		<SolidColorBrush x:Key="FilledTextBoxDeleteButtonForegroundDisabled"
						 Color="Blue" />

		<SolidColorBrush x:Key="FilledTextBoxForeground"
						 Color="Indigo" />
		<SolidColorBrush x:Key="FilledTextBoxForegroundPointerOver"
						 Color="Pink" />
		<SolidColorBrush x:Key="FilledTextBoxForegroundFocused"
						 Color="Red" />
		<SolidColorBrush x:Key="FilledTextBoxForegroundDisabled"
						 Color="Orange" />

		<SolidColorBrush x:Key="FilledTextBoxPlaceholderForeground"
						 Color="Yellow" />
		<SolidColorBrush x:Key="FilledTextBoxPlaceholderForegroundPointerOver"
						 Color="Green" />
		<SolidColorBrush x:Key="FilledTextBoxPlaceholderForegroundFocused"
						 Color="Blue" />
		<SolidColorBrush x:Key="FilledTextBoxPlaceholderForegroundDisabled"
						 Color="Indigo" />
		

		
		<SolidColorBrush x:Key="OutlinedTextBoxBorderBrush"
						 Color="Pink" />
		<SolidColorBrush x:Key="OutlinedTextBoxBorderBrushPointerOver"
						 Color="Red" />
		<SolidColorBrush x:Key="OutlinedTextBoxBorderBrushFocused"
						 Color="Orange" />
		<SolidColorBrush x:Key="OutlinedTextBoxBorderBrushDisabled"
						 Color="Yellow" />

		<SolidColorBrush x:Key="OutlinedTextBoxForeground"
						 Color="Green" />
		<SolidColorBrush x:Key="OutlinedTextBoxForegroundPointerOver"
						 Color="Blue" />
		<SolidColorBrush x:Key="OutlinedTextBoxForegroundFocused"
						 Color="Indigo" />
		<SolidColorBrush x:Key="OutlinedTextBoxForegroundDisabled"
						 Color="Pink" />

		<SolidColorBrush x:Key="OutlinedTextBoxPlaceholderForeground"
						 Color="Red" />
		<SolidColorBrush x:Key="OutlinedTextBoxPlaceholderForegroundPointerOver"
						 Color="Orange" />
		<SolidColorBrush x:Key="OutlinedTextBoxPlaceholderForegroundFocused"
						 Color="Yellow" />
		<SolidColorBrush x:Key="OutlinedTextBoxPlaceholderForegroundDisabled"
						 Color="Green" />
		
	</Page.Resources>

## PR Checklist 
- Commits must be following the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [X] Tested UWP
- [X] Tested iOS
- [X] Tested Android
- [X] Tested WASM
- [ ] Tested MacOS
- [X] Contains **No** breaking changes

## Internal Issue (If applicable):

https://github.com/unoplatform/Uno.Themes/issues/988

### CODE REVIEW

File: src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml [hunk 1]
```diff
@@ -13,17 +13,192 @@
...
+			<!--#region Material Filled Text Box Style-->
+			<StaticResource x:Key="MaterialFilledTextBoxForeground"
+							ResourceKey="OnSurfaceBrush" />
+			<StaticResource x:Key="MaterialFilledTextBoxPlaceholderForeground"
+							ResourceKey="OnSurfaceVariantBrush" />
+			<StaticResource x:Key="MaterialFilledTextBoxBackground"
+							ResourceKey="SurfaceVariantBrush" />
+			<StaticResource x:Key="MaterialFilledTextBoxBorderBrush"
+							ResourceKey="OnSurfaceVariantBrush" />
+			<StaticResource x:Key="MaterialFilledTextBoxPointerOverBorderBrush"
```
Comment (silviuo) [inline src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml:37]: Hey Vy, my understanding was that the visual state comes last in the naming convention: 
{design}{?:style-variant}{control-name}{member}{?:visual-state}

So in this case `MaterialFilledTextBoxPointerOverBorderBrush` would become `MaterialFilledTextBoxBorderBrushPointerOver`

Comment (nguyeenvy) [reply src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml:37]: @silviuo I will fix that! Thanks :)

File: src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml [hunk 2]
```diff
@@ -300,8 +300,8 @@
 			<SolidColorBrush x:Key="OnSurfaceDraggedBrush" Color="{StaticResource OnSurfaceColor}" Opacity="{StaticResource DraggedOpacity}" />
 			<SolidColorBrush x:Key="OnSurfaceSelectedBrush" Color="{StaticResource OnSurfaceColor}" Opacity="{StaticResource SelectedOpacity}" />
 			<SolidColorBrush x:Key="OnSurfaceMediumBrush" Color="{StaticResource OnSurfaceColor}" Opacity="{StaticResource MediumOpacity}" />
-			<SolidColorBrush x:Key="OnSurfaceLowBrush" Color="{StaticResource OnSurfaceColor}" Opacity="{StaticResource LowOpacity}" />
-			<SolidColorBrush x:Key="OnSurfaceDisabledBrush" Color="{StaticResource OnSurfaceColor}" Opacity="{StaticResource DisabledOpacity}" />
+			<SolidColorBrush x:Key="OnSurfaceLowBrush" Color="{StaticResource OnSurfaceColor}" Opacity="{StaticResource DisabledOpacity}" />
+			<SolidColorBrush x:Key="OnSurfaceDisabledBrush" Color="{StaticResource OnSurfaceColor}" Opacity="{StaticResource HoverOpacity}" />
```
Comment (kazo0) [inline src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml:304]: These should be kept as `LowOpacity` and `DisabledOpacity`, no?

File: src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml [hunk 3]
```diff
@@ -13,17 +13,134 @@
 										NullOrEmptyValue="0" />
 
 	<!-- Path Data -->
-	<x:String x:Key="M3ClearGlyphPathData">M10 0C4.47 0 0 4.47 0 10C0 15.53 4.47 20 10 20C15.53 20 20 15.53 20 10C20 4.47 15.53 0 10 0ZM10 18C5.59 18 2 14.41 2 10C2 5.59 5.59 2 10 2C14.41 2 18 5.59 18 10C18 14.41 14.41 18 10 18ZM10 8.59L13.59 5L15 6.41L11.41 10L15 13.59L13.59 15L10 11.41L6.41 15L5 13.59L8.59 10L5 6.41L6.41 5L10 8.59Z</x:String>
+	<x:String x:Key="MaterialTextBoxClearGlyphData">M14.9482 6.46442L13.534 5.05021L9.99849 8.58574L6.46296 5.05021L5.04874 6.46442L8.58428 9.99995L5.04874 13.5355L6.46296 14.9497L9.99849 11.4142L13.534 14.9497L14.9482 13.5355L11.4127 9.99995L14.9482 6.46442ZM17.0696 2.92889C13.1663 -0.974342 6.83065 -0.974342 2.92742 2.92889C-0.975807 6.83212 -0.975807 13.1678 2.92742 17.071C6.83065 20.9743 13.1663 20.9743 17.0696 17.071C20.9728 13.1678 20.9728 6.83212 17.0696 2.92889ZM4.34164 15.6568C1.22329 12.5385 1.22329 7.46144 4.34164 4.3431C7.45998 1.22476 12.537 1.22476 15.6553 4.3431C18.7737 7.46144 18.7737 12.5385 15.6553 15.6568C12.537 18.7752 7.45998 18.7752 4.34164 15.6568Z</x:String>
```
Comment (kazo0) [inline src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml:16]: Is this new path data taken from the Figma file, did the old one not match?

File: src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml [hunk 4]
```diff
@@ -13,17 +13,134 @@
...
 
 	<SolidColorBrush x:Key="MaterialDisabledOutlinedTextBoxBorderBrush"
 					 Opacity="0.12"
 					 Color="{ThemeResource OnSurfaceColor}" />
 
+	<ResourceDictionary.ThemeDictionaries>
```
Comment (kazo0) [inline src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml:22]: Let's keep the `ThemeDictionaries` declaration at the top of the file and move the non-theme resources below it

File: src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml [hunk 5]
```diff
@@ -13,17 +13,134 @@
...
+			<StaticResource x:Key="FilledTextBoxBorderBrushPointerOver" ResourceKey="OnSurfaceBrush" />
+			<StaticResource x:Key="FilledTextBoxBackgroundPointerOver" ResourceKey="OnSurfaceVariantHoverBrush" />
+			<StaticResource x:Key="FilledTextBoxBorderBrushDisabled" ResourceKey="OnSurfaceDisabledBrush" />
+			<StaticResource x:Key="FilledTextBoxForegroundDisabled" ResourceKey="OnSurfaceDisabledBrush" />
+			<StaticResource x:Key="FilledTextBoxPlaceholderForegroundDisabled" ResourceKey="OnSurfaceLowBrush" />
+			<StaticResource x:Key="FilledTextBoxBackgroundDisabled" ResourceKey="OnSurfaceDisabledBrush" />
+			<StaticResource x:Key="FilledTextBoxDeleteButtonForegroundDisabled" ResourceKey="OnSurfaceLowBrush" />
+			<StaticResource x:Key="FilledTextBoxPlaceholderForegroundFocused" ResourceKey="PrimaryBrush" />
+			<StaticResource x:Key="FilledTextBoxBorderBrushFocused" ResourceKey="PrimaryBrush" />
+			<!--#endregion-->
```
Comment (kazo0) [inline src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml:41]: Let's organize the brushes similar to the [Fluent theme resources](https://github.com/unoplatform/uno/blob/b4673c31ae536352e82ccc1dcebf221f3debd0d0/src/Uno.UI.FluentTheme.v2/Resources/Version2/Priority02/TextBox_themeresources.xaml#L32-L48) where all Background + Visual State brushes are clumped together, and then all Foreground + Visual State Brushes are grouped together, etc. 

This way we can easily scan the list to find the appropriate brush based on the property being set

File: src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml [hunk 6]
```diff
@@ -113,9 +230,9 @@
 								<VisualState x:Name="PointerOver">
 									<VisualState.Setters>
 										<Setter Target="NormalBorder.Fill"
-												Value="{ThemeResource OnSurfaceBrush}" />
-										<Setter Target="NormalBorder.Height"
-												Value="2" />
```
Comment (kazo0) [inline src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml:118]: Should we be removing the NormalBorder.Height setter here?

File: src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml [hunk 7]
```diff
@@ -347,42 +466,39 @@
...
 							<VisualStateGroup x:Name="CommonStates">
 								<VisualState x:Name="Normal" />
 								<VisualState x:Name="PointerOver">
 									<VisualState.Setters>
 										<Setter Target="RootBorder.BorderBrush"
-												Value="{ThemeResource OnSurfaceBrush}" />
-										<Setter Target="RootBorder.BorderThickness"
-												Value="2" />
-										<Setter Target="RootBorder.Padding"
-												Value="0" />
```
Comment (kazo0) [inline src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml:362]: These setters should stay, no?

File: src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml
Comment (kazo0) [inline src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml:1]: We have some gaps here. Looking at the [Fluent theme resources](https://github.com/unoplatform/uno/blob/b4673c31ae536352e82ccc1dcebf221f3debd0d0/src/Uno.UI.FluentTheme.v2/Resources/Version2/Priority02/TextBox_themeresources.xaml#L32-L47C121), we should at least have a themed resource for each of the properties: Background/Foreground/BorderBrush/PlaceholderForeground, for each visual state: Normal(blank)/PointerOver/Focused/Disabled

Even if things like Foreground don't actually change color between different visual states, a Brush alias should still be defined and set in the proper Visual State. That way, if someone wanted to customize per visual state, they wouldn't need to re-template

File: src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml (hunk 2 above)
Comment (nguyeenvy) [reply src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml:304]: @kazo0 On Figma, I see that OnSurfaceDisabledBrush is set to 8% opacity which is HoverOpacity, and OnSurfaceLowBrush is set to 38% opacity which is DisabledOpacity. That is why I changed them, but I can always revert these changes if you want to keep them as is.

image
image

File: src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml (hunk 3 above)
Comment (nguyeenvy) [reply src/library/Uno.Material/Styles/Controls/v2/TextBox.xaml:16]: @kazo0 Yes it is from the Figma file. The old one looked a bit different from the one on Figma (smaller and less thick). Do you want to revert to the old path?

File: src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml (hunk 2 above)
Comment (kazo0) [reply src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml:304]: @NVLudwig is there a mistake in our code or yours? :)

Comment (kazo0) [reply src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml:304]: Is this maybe a larger problem, are all "Low" brushes supposed to be at 38% and all "Disabled" supposed to be at 8% and not just the OnSurface brushes?

If that's the case should we maybe just be changing the base values for LowOpacity and DisabledOpacity?

**Review by kazo0 — APPROVED** [review]

File: src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml (hunk 2 above)
Comment (kazo0) [reply src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml:304]: @nguyeenvy for now lets leave the brushes with the LowOpacity and DisabledOpacity just so it stays consistent with naming and then we can deal with changing their values in another issue

**Pull request merged**